### PR TITLE
NIFI-4142: Refactored Record Reader/Writer to allow for reading/writi…

### PR DESCRIPTION
**Pull request by markap14**
…ng "raw records". Implemented ValidateRecord.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/SchemaValidationException.java [hunk 1]
```diff
@@ -15,14 +15,16 @@
  * limitations under the License.
  */
 
-package org.apache.nifi.serialization.record;
+package org.apache.nifi.serialization;
 
-public class TypeMismatchException extends RuntimeException {
```
Comment (joewitt) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/SchemaValidationException.java]: i dont think it is ok to change this exeception class name at this juncture and even if it is questionable ok the juice is probably not worth the squeeze.  TypeMismatch and SchemaValidation are pretty much the same thing

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java [hunk 2]
```diff
@@ -38,14 +38,35 @@
...
+     * @param enforceSchema whether or not fields in the Record should be validated against the schema and coerced when necessary
+     *
+     * @return the next record in the stream or <code>null</code> if no more records are available
+     * @throws IOException if unable to read from the underlying data
+     * @throws MalformedRecordException if an unrecoverable failure occurs when trying to parse a record, or a Record contains a field
+     *             that violates the schema and cannot be coerced into the appropriate field type.
+     * @throws SchemaValidationException if a Record contains a field that violates the schema and cannot be coerced into the appropriate
+     *             field type and schema enforcement is enabled
+     */
+    Record nextRecord(boolean enforceSchema) throws IOException, MalformedRecordException;
```
Comment (joewitt) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java]: the schema had always been enforced arguably just with sense of leniency.  I think this method parameter should be 'strictSchemaEnforcement' or 'enforceStrictSchema'.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java [hunk 3]
```diff
@@ -38,14 +38,35 @@
...
+     * Returns the next record in the stream or <code>null</code> if no more records are available. Schema enforcement will be enabled.
      *
      * @return the next record in the stream or <code>null</code> if no more records are available.
      *
      * @throws IOException if unable to read from the underlying data
      * @throws MalformedRecordException if an unrecoverable failure occurs when trying to parse a record
+     * @throws SchemaValidationException if a Record contains a field that violates the schema and cannot be coerced into the appropriate field type.
      */
-    Record nextRecord() throws IOException, MalformedRecordException;
+    default Record nextRecord() throws IOException, MalformedRecordException {
```
Comment (joewitt) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java]: we should indicate whether the scheme enforcement strictness is 'lenient' or 'strict'.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/SchemaValidationException.java (hunk 1 above)
Comment (markap14) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/SchemaValidationException.java]: I don't agree that they are pretty much the same thing. TypeMismatchException is very specific. SchemaValidationException can be much more broad. For instance, if a required field is missing, that is not a Type Mismatch, but it is a Schema Validation.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java (hunk 2 above)
Comment (markap14) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/RecordReader.java]: I think I actually want to just separate the concept out into two different variables here: boolean coerceTypes, boolean dropUnknownRecords. That way it is very explicit what is happening, and I don't think that 'strict' vs. 'lenient' really conveys those two semantics as well as I'd like.

**Comment by markap14** [comment]
@joewitt I've pushed a new commit that I believe better clarifies how schemas are treated in terms of strictness vs. leniency by providing two arguments instead of 'enforceSchema': 'coerceTypes' and 'dropUnknownFields'

**Comment by joewitt** [comment]
@markap14 can you please rebase and resolve conflict and please go ahead and squash.

**Comment by joewitt** [comment]
@markap14  looks like the option says "Used Reader's Schema".  I'm assuming that is a typo and it should say "Use Reader's Schema".  But that is not in this PR it appears.  Did that sneak in elsewhere?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateRecord.java [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    + "FlowFiles if some records are valid according to the schema and others are not. Any FlowFile that is routed to the \"invalid\" relationship will emit a ROUTE Provenance Event "
+    + "with the Details field populated to explain why records were invalid. In addition, to gain further explanation of why records were invalid, DEBUG-level logging can be enabled "
+    + "for the \"org.apache.nifi.processors.standard.ValidateRecord\" logger.")
+public class ValidateRecord extends AbstractProcessor {
+
+    static final AllowableValue SCHEMA_NAME_PROPERTY = new AllowableValue("schema-name-property", "Use Schema Name Property",
+        "The schema to validate the data against is determined by looking at the 'Schema Name' Property and looking up the schema in the configured Schema Registry");
+    static final AllowableValue SCHEMA_TEXT_PROPERTY = new AllowableValue("schema-text-property", "Use Schema Text Property",
+        "The schema to validate the data against is determined by looking at the 'Schema Text' Property and parsing the schema as an Avro schema");
+    static final AllowableValue READER_SCHEMA = new AllowableValue("reader-schema", "Used Reader's Schema",
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateRecord.java]: Ah here it is.  "Used Reader's Schema" should be "Use Reader Schema".

**Comment by joewitt** [comment]
went ahead and fixed the typo and squashed locally.

Ran a test flow that created csv of various conformity to a simple spec.  Ran it through various configratuions of the validator and it behaved as desired and expected in each case.  When data was marked invalid was able to see precisely why through the provenance events which was great.  Was able to see coercion/non-coercion behavior and all checks out.

+1 merging to master

**Pull request closed without merging**